### DISCUSSION TITLE
Future of the Project

### DISCUSSION THREAD

**Discussion by Zerwin** [bttv-android/bttv#669] [General]
It seems like the originals authors (@FoseFx ) interest has faded in this project, or maybe they just are too busy with other things in life. I tried my hand at building the app (which worked for 15.4.1) and slowly incrementally updating the base version, but I have never worked with git patches before and ran into several issues that stopped me. 

Judging from all the open issues and discussion about updates it definitely seems like there is an interest from ppl to have a working version of this app, however, until now nobody has stepped forward.

---

### My question would be, are there any other ppl that want to maintain this project ?

Maybe FoseFx could also just provide a small update if they want to work on this project again at some point in the future or if they would be fine with transferring it to someone else.

I would also like to gauge if there is anyone who could provide help with just patching/reverse engineering to newer base versions ?

The main issue I am now running into is the creation of patches. I ran the update script in the repository, solved the merge conflicts as good as I could, but when I ran the script to create monkey patches it outputted several thousand files. I seem to have made an error somewhere along the line but currently can't figure out where or what it was.

**Comment by FoseFx** [comment]
Thank you everyone for your continued interest in this project.

Unfortunately, rebasing the app requires more time than I currently have available.
I used to test every release for some time before I'd publish it, but now I pretty much never use the Twitch platform anymore, so I need to add even more time on top of that.

I have written a bit of build documentation here:
- https://github.com/bttv-android/bttv?tab=readme-ov-file#build-it-yourself
- https://blog.bmn.dev/I-wrote-an-open-source-mod-of-an-Android-App
- https://github.com/bttv-android/bttv/blob/master/CONTRIBUTING.md#setup
- https://github.com/bttv-android/bttv/blob/master/architecture.md
- https://github.com/bttv-android/bttv/wiki/Checklists#update

Required knowledge:
- General bash/linux knowledge
- git
- Java, Android SDK
- Smali (you'll figure this out with time, at least that's what I did)

If someone is willing to invest the time, I'll be happy to help with any questions.

Some tools, like the ubi script, probably still have hard coded paths, so you might need to update them

**Comment by FoseFx** [comment]
> The main issue I am now running into is the creation of patches. I ran the update script in the repository, solved the merge conflicts as good as I could, but when I ran the script to create monkey patches it outputted several thousand files. I seem to have made an error somewhere along the line but currently can't figure out where or what it was.

I'll be happy to provide you with some assistance with this, you can, for example, contact me over email (you can find my address on my website).

**Reply by Zerwin** [reply]
In the vein that maybe this will also help someone else I would keep my questions here on Github for now, but thanks for the offer via Email already.

So I followed the checklist as good as I could, I provided the original 15.4.1 as old.apk (this is the same APK I did manage to build a working version with) and then provided 15.6.2 as new APK ( wanted to start with little steps and saw you already had a branch for this version some time ago).

I then ran the update.sh script, and it finished till the end where it aborts the merge. I tried solving the merge problems as best as I could (not sure what was meant with "Change rename limit"). But this is basically where I get stuck at, I'm not sure what's meant with generating new id's, and running the genmonke script just deletes all the existing patches and produces a few thousand new ones. This seems like a bit too many patches to be right.

Any idea what I'm doing wrong here ?

**Reply by FoseFx** [reply]
By default the git rename algorithm has a limit for performance reasons.

Because there are a lot of limes which move around, this limit is reached.

It should give you a recommendation on what to set this limit to in the warning.

Use 

```
git config merge.renameLimit <high value>
```
in the inner repo, or alternatively set it globally (`git config --global merge...`) 

And retry the merge part of the update script.

**Reply by Zerwin** [reply]
Okay, I set that option, redid the merge and solved the merge conflicts. What should be the next step ?

**Reply by FoseFx** [reply]
[The next step](https://github.com/bttv-android/bttv/wiki/Checklists#update) is to use the [build-companion](https://github.com/bttv-android/build-companion/) to generate new ids for the resources we introduce to the apk. Use `build-companion ids <path to the public.xml, should be in res of the disass directory>`.

**Reply by Zerwin** [reply]
Okay, if I didn't get any output I can assume build-companion was successful ? As for generate monke.patch, is that just the genmmonke script ?

**Reply by FoseFx** [reply]
> Okay, if I didn't get any output I can assume build-companion was successful

I think so. You can also check if the public.xml file has changed, now all ids for resources (bttv_ prefix) should start where the values before it have ended.

> Yes, in previous versions only a single patch file was generated. Now the genmonke script generates one for every file, and orders them according to the feature that was patched in in this file. It might happen, that smali files were renamed, in that case you should update the genmonke script, so keep an eye on the console output.

**Reply by FoseFx** [reply]
I forgot to mention, you need to run the genmonke script with `BASE=update ./genmonke`

**Reply by Zerwin** [reply]
Okay, that did it, without BASE=update I got several thousand patches, but with it I only got around 20, which seems more sane.

**Reply by Zerwin** [reply]
Okay, so I managed to get the app updated to 15.9.0 without much pain (this version apparently already had the featured clips feature). I noticed a few bugs but before I even get to taking a look at these, I'm more stuck on how to figure out some of the patches.
Like I saw that you had to fix the login at some point in the past https://github.com/bttv-android/bttv/pull/589/commits/52262cc54c32bc060ad7e51c50b9375ea4bfbdf3 , but I'm kinda stumped, how did you figure out you had to change exactly these lines of code to make the login work ?

**Reply by FoseFx** [reply]
Welcome to the dark art of reverse engineering.

In that particular case identifying the root cause actually took a lot of time, as there were no code changes responsible for the issue. I had to intercept the network traffic, I used mitm-proxy with a custom certificate root and a networking patch. If I had to do it today, I'd use something like Frida to hook the app and go from there.

With the [root cause](https://github.com/bttv-android/bttv/issues/588#issuecomment-1440073772) identified, I just needed to know where the User-Agent was stored. For that I looked for usages of typical setup functions of the http library that they used in the decompiled code base.
Once I knew what function needed to be patched, I read the smali class definition of that function and cross-referenced it with the decompiled version. I then knew which register helt the string, and then simply added a statement that overwrites it with a static string (the one the backend expected).

Some of this workflow is explained [here](https://blog.bmn.dev/I-wrote-an-open-source-mod-of-an-Android-App).

**Reply by Zerwin** [reply]
I seem to run into https://github.com/iBotPeaches/Apktool/issues/2870 this issue more and more often now, the build fails with

```
Exception in thread "main" org.jf.util.ExceptionWithContext: Exception occurred while writing code_item for method Ltv/twitch/android/shared/chat/ChatDisconnectTiming;->values()[Ltv/twitch/android/shared/chat/ChatDisconnectTiming;
        at org.jf.dexlib2.writer.DexWriter.writeDebugAndCodeItems(DexWriter.java:1046)
        at org.jf.dexlib2.writer.DexWriter.writeTo(DexWriter.java:345)
        at org.jf.dexlib2.writer.DexWriter.writeTo(DexWriter.java:300)
        at brut.androlib.src.SmaliBuilder.build(SmaliBuilder.java:61)
        at brut.androlib.src.SmaliBuilder.build(SmaliBuilder.java:36)
        at brut.androlib.Androlib.buildSourcesSmali(Androlib.java:420)
        at brut.androlib.Androlib.buildNonDefaultSources(Androlib.java:366)
        at brut.androlib.Androlib.build(Androlib.java:304)
        at brut.androlib.Androlib.build(Androlib.java:270)
        at brut.apktool.Main.cmdBuild(Main.java:259)
        at brut.apktool.Main.main(Main.java:85)
Caused by: org.jf.util.ExceptionWithContext: Error while writing instruction at code offset 0x2
        at org.jf.dexlib2.writer.DexWriter.writeCodeItem(DexWriter.java:1319)
        at org.jf.dexlib2.writer.DexWriter.writeDebugAndCodeItems(DexWriter.java:1042)
        ... 10 more
Caused by: org.jf.util.ExceptionWithContext: Unsigned short value out of range: 65538
        at org.jf.dexlib2.writer.DexDataWriter.writeUshort(DexDataWriter.java:116)
        at org.jf.dexlib2.writer.InstructionWriter.write(InstructionWriter.java:356)
        at org.jf.dexlib2.writer.DexWriter.writeCodeItem(DexWriter.java:1279)
        ... 11 more
```
It's not always specifically ChatDisconnectTiming but rather some random function/file. Did you have the same problem back then and maybe found a solution ?

**Reply by FoseFx** [reply]
The issue here is that the smali classes become too big to be compiled into a dex file. The fix is to simply move some files out of the offending smali directory. I assume this will worsen either bundle size, performance, or both. Try to comment one of these back in:
https://github.com/bttv-android/bttv/blob/master/disassemble#L147-L156

**Reply by FoseFx** [reply]
From my testing, on the latest 15.x release I was not able to log in using the unmodded app, so if you are trying to debug anything there, it might just be that the backend apis the app uses were deprecated.

**Reply by Zerwin** [reply]
I couldn't login with any release in the emulator, but on my real phone it worked so I just use that for testing.

**Reply by Zerwin** [reply]
I think the login issues on emulators is something Twitch enabled/hardened. On Login in an emulator, I can see a POST call to https://passport.twitch.tv/integrity, which returns a token. The token is then part of the next POST call to https://passport.twitch.tv/protected_login, which returns with error code 400, Bad Request and the following body:
```
{
    "error": "Oops! We encountered an unexpected error. Please try again.",
    "error_code": 5021,
    "error_description": "invalid integrity token"
}
```

I'm guessing they are doing some sort of verification there and expecting something different in the second call which fails. Will try and see if I can get a functioning integrity token as an example somewhere.

**Reply by FoseFx** [reply]
They are using https://www.kasada.io/, it's a deep rabbit hole if you want to reverse that. They seem to send a JavaScript script which has some kind of obfuscated turing machine. Other than that the initial integrity request contains "encrypted" (key is hardcoded) device information.

**Reply by Zerwin** [reply]
Yeah, I tried looking into a bit more but couldn't even get a valid response when copying the exact same requests from my browser console in a REST API tool for debugging. As this should only affect developing and not real device usage I'll guess we just have to live with it though.

**Reply by Zerwin** [reply]
Okay, funny workaround, Waydroid Emulator works fine and allows me to login.

**Reply by Zerwin** [reply]
Is there anything speaking against commiting the original Twitch APK file ? An APKMirror URL doesn't work in CI (probably cause APKMirror blocks it).

**Reply by Zerwin** [reply]
So good news, yay, I got a build of 15.7.0 done in the Gitlab CI. https://github.com/Zerwin/bttv/actions/runs/10343069602 I had to change the CI script to make it run without the docker container because I needed to move some of the smali folders in one of the scripts and the docker image uses the scripts it had at build time. I also had to use an APKMirror url that will probably expire soon, but other than that it's looking good. 

I picked 15.7.0 because that is the version where they put clips up at the top of the Home page, meaning there's some actual change to see.

I would like to have that version out and then see if anyone reports any bugs (other than the already known ones), and then just slowly upgrade a few versions every two weeks or so until we mostly catch up.

After we figure out a way to get a proper URL for the APKMirror URL, or if commiting them would also be fine, I would do a PR and we can merge the changes and roll the update out (I saw there was something related to auto updates of BTTV, but didn't look further into how it works yet)

**Reply by Zerwin** [reply]
Got until 16.0 https://github.com/Zerwin/bttv/actions/runs/10375901375 with simple updates, then I encountered some bug I have yet to take a closer look at.

**Reply by Zerwin** [reply]
I mostly figured out how to deal with issues during updates now. My latest branch compiled a functioning 16.9 version, which includes the new features of stories, watch streaks and highlighted clips. I think this would make a good point for a full release. @FoseFx Do you think we could merge this during the weekend ?

**Reply by FoseFx** [reply]
Thank you for your work! Please open a PR, then I'll take a look!

**Comment by Zerwin** [comment]
From what I can see, the two main problems the app currently has are:

- Old baseline
- Bugs with BTTV features

I'm a bit undecided on what should be the focus first, as both problems kinda work against each other. 

Rebasing to a newer baseline will probably introduce more bugs into existing features, but allow for more of the new twitch app features to be used. This should probably be done incrementally, so from the current 15.X to 16.0 to 17.0 ... until 20.X (but stop before the new twitch layout was introduced, because avoiding that is basically the entire reason I got motivation to get involved in this project).

Fixing the bugs would be good, but the rebasing will break these probably over and over, so maybe ignoring them for now and focusing on the rebasing would be a good idea ? A bit undecided about that.

**Reply by FoseFx** [reply]
I think the main feature bttv/ffz/7tv emotes must work, everything else may be unstable, as long as one can keep them disabled without problems. At this point, I think it is ok to even remove features that cause issues, and patch them in later again.

**Comment by thehandsomedan** [comment]
Thank you for stepping in to offer assistance @Zerwin , it is very much appreciated. Do you have any plans for future updates to the app?

**Reply by Zerwin** [reply]
Currently no, most of the changes of this project only work up to 16.9, after that Twitch did a lot of internal refactoring it seems that makes the patches not apply anymore. By now there are so many changes between 16.9 and the current version that starting over would probably be easier.
